bittensor/_dataset/dataset_impl.py

```diff
@@ -1,678 +1,750 @@
+            loop (asyncio.Loop):
+                Asyncio loop 
+        """
+        if datasets == None:
+            datasets = self.datasets
 
-        self.build_hash_table()
+        all_text_file_metas = []
+        dataset_hash_map = {}
 
-        os.makedirs(os.path.expanduser(data_dir), exist_ok=True)
-            
-        self.data_queue = ThreadQueue(
-            producer_target = self.reserve_multiple_data,
-            producer_arg = (self.num_batches, ),
-            buffer_size = 1
-        )
+        tasks = []
 
-    def __del__(self):
-        self.close()
+        self.dataset_size_map = {d:0 for d in datasets}
+        
+        # Gather dataset hashes async as their state is independent.
+        for dataset in datasets:
+            tasks += [self.async_build_single_dataset(dataset=dataset, save=save, load=load)]
 
-    def close(self):
-        self.data_queue.close()
+        # Get the hashes asynchronously for each dataset.
+        dataset_hashes = asyncio.run(asyncio.gather(*tasks))
 
-    def get_folder_size(self, folder):
-        r""" Get the size (in byte) of a folder inside the data_dir.
-        Args:
-            folder (str):
-                The name of the folder
+        # Create a hash map of dataset -> text hashes.
+        for k,v in zip(datasets, dataset_hashes):
+            if len(v) > 0:
+                dataset_hash_map[k] = v
         
-        Returns:
-            total_size (int):
-                The memory size of the folder (in byte). 
-        """
-        total_size = 0
-        full_path = os.path.expanduser(os.path.join(self.data_dir, folder))
-        for dirpath, dirnames, filenames in os.walk(full_path):
-            for f in filenames:
-                fp = os.path.join(dirpath, f)
-                # skip if it is symbolic link
-                if not os.path.islink(fp):
-                    total_size += os.path.getsize(fp)
-
-        return total_size
-
-    def load_hash(self, file_meta):
-        r""" Load a hash from disk.
-        Args:
-            file_meta (dict of str: int):
-                Specify the details of the dataset in the format of {'Name': , 'Hash':}.
+        self.dataset_hash_map = dataset_hash_map
+
+        # Flatten the hashes to a list of hashes.
+        for  k,file_meta_list in dataset_hash_map.items():
+           
+            all_text_file_metas += [fm  for fm in file_meta_list if fm['Size'] >= self.block_size_bytes]
+        assert len(all_text_file_metas) > 0
+        self.all_text_file_metas = all_text_file_metas
+
+    async def async_save_json(self, 
+                              path:str,
+                              obj:Union[dict, list],
+                              include_root:bool=True) -> str:
+        """ 
+        Async save of json for storing text hashes
 
-        Returns:
-            text (str): 
-                The text in the file.                
+        Args:
+            path (List[str]):
+                Axon to serve.
```

Comment:
Again.

Reply:
did a run through lingering syntax to make her pretty for the prom.